File: catkin_ws/src/students/sandoval_penilla/scripts/practica_02.py
#!/usr/bin/env python
# coding=utf-8

import roslib
roslib.load_manifest('sandoval_penilla')
import rospy
import math
import tf
import sys
from geometry_msgs.msg import Twist

NOMBRE = "sandoval_penilla"

class Practica02:
    """
    ==============================================
    Practica_02 Sandoval Penilla Oscar
    UNAM Facultad de Ingenieria
    Temas Selectos de MEcatronica
    
       
    ==============================================
    """
    
    def __init__(self):
        print("PRÀCTICA 02 - Control de Posición - {}".format(NOMBRE))
        self.n = rospy.init_node("practica_02")
        self.run = False
        
        # Obtencion de parametros
        self.ag = 0.0 # 0.0 rad por defecto
        self.type = "diff" #metodo diff holonomico por defecto
        
        # validacion de parametros        
        if rospy.has_param("~x") and rospy.has_param("~y"):
            self.xg = rospy.get_param("~x")
            self.yg = rospy.get_param("~y")      
            self.run = True
        else: print("parmeters X and Y are mandatory")
           
        if rospy.has_param("~type"):
            if rospy.get_param("~type") == "diff" or rospy.get_param("~type") == "omni":
                self.type = rospy.get_param("~type")
            else:
                print("Valid control types are \omni\ and \diff\ ")
                self.run = False
        
        if rospy.has_param("~a"):
            self.ag = rospy.get_param("~a")
            
        # Borramos los parametros para una nueva ejecucion
        try:
            rospy.delete_param("~x")
            rospy.delete_param("~y")
            rospy.delete_param("~a")
            rospy.delete_param("~type")
        except KeyError:
            pass
            
        
        # Creamos nuestro mensaje de tipo twist        
        self.m = Twist()
        # Creamos nuestro listener de transformaciones
        self.listener = tf.TransformListener()
        # Creamos un publisher para publicar el mensaje twist
        self.publisher = rospy.Publisher('/hardware/mobile_base/cmd_vel',
                                  Twist,
                                  queue_size=1)
                
    def diferencial(self):
        print("Executing position control for an DIFFERENTIAL base with \n Goal X={} \tGoal Y={} \tGoal Angle={}".format(self.xg,self.yg,self.ag))            
   
        rate = rospy.Rate(10)
        while not rospy.is_shutdown():
            try:
                self.listener.waitForTransform("map", "base_link", 
                                          rospy.Time(), 
                                          rospy.Duration(1.0))
                (trans,rot) = self.listener.lookupTransform('map',
                                                            'base_link',
                                                             rospy.Time(0))
                tf = self.listener.lookupTransform('map',
                                                            'base_link',
                                                             rospy.Time(0))
                
            except (tf.LookupException, tf.ConnectivityException, 
                    tf.ExtrapolationException):
                    pass
                
            goal = False
            xr = trans[0]
            yr = trans[1]
            v_max = 0.5
            a = 1
            if not goal:
                thr = math.atan2(rot[2],rot[3])*2
                eth = math.atan2(self.yg - yr, self.xg - xr) - thr
                #distancia al punto
                dist = math.sqrt((self.yg - yr)**2 + (self.xg - xr)**2)
                if eth < -math.pi: eth += 2*math.pi
                elif eth > math.pi: eth += -2*math.pi
    
                b = 1
                w_max = 1   
                v = v_max*math.exp((-eth**2)/a)
                w = w_max*(((2)/(1+math.exp(-eth/b)))-1)
                
                self.m.angular.z = w
                self.m.linear.x = v
                self.publisher.publish(self.m)

            # el robot esta dentro de la tolerancia po lo que se dejara de avanzar
            if dist < 0.1:
                goal = True
                self.m.linear.x = 0
                self.publisher.publish(self.m)

            rate.sleep()
   
    def omni(self):
        print("Executing position control for an OMNIDIRECTIONAL base with \n Goal X={} \tGoal Y={} \tGoal Angle={}".format(self.xg,self.yg,self.ag))            
        rate = rospy.Rate(10)
        while not rospy.is_shutdown():
            try:
                self.listener.waitForTransform("map", "base_link", 
                                          rospy.Time(), 
                                          rospy.Duration(1.0))
                (trans, rot) = self.listener.lookupTransform('map',
                                                            'base_link',
                                                             rospy.Time(0))
            except (tf.LookupException, tf.ConnectivityException, 
                    tf.ExtrapolationException):
                pass
            
            kp = 1            
            xr = trans[0]
            yr = trans[1]
            thr = math.atan2(rot[2],rot[3])*2
            
            ex = self.xg - xr
            ey = self.yg - yr
            ea = self.ag - thr
            
            
            vx = kp*ex
            vy = kp*ey
            w  = kp*ea
            # controlamos la velocidad max, se reduce en un 10% la velocidad
            while math.fabs(vx) > 0.5 or math.fabs(vy) > 0.5:
                vx *= 0.9
                vy *= 0.9
            
            self.m.linear.x = vx
            self.m.linear.y = vy
            self.m.angular.z = w
            self.publisher.publish(self.m)
            
            rate.sleep()
            
            
    def main(self):
        if self.run and self.type == "diff":
            self.diferencial()
        elif self.run and self.type == "omni":
            self.omni()


if __name__ == '__main__':
    
    try:
        prac = Practica02()
        prac.main()
        
    except rospy.ROSInterruptException:
        pass

File: catkin_ws/src/students/sandoval_penilla/src/Practica02.cpp
/*
 * ROBOTS MÓVILES Y AGENTES INTELIGENTES
 * FACULTAD DE INGENIERÍA, UNAM, 2019-1
 * P R Á C T I C A   2
 * CONTROL DE POSICIÓN
 */

#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "tf/transform_listener.h"
#define NOMBRE "Sandoval_Penilla"

int main(int argc, char** argv)
{
    std::cout << "PRÁCTICA 02 - Control de Posición - " << NOMBRE << std::endl;
    ros::init(argc, argv, "practica_02");
    ros::NodeHandle n("~");

    if(!n.hasParam("x") || !n.hasParam("y"))
    {
	std::cout << "Parameters \"x\" and \"y\" are required. " << std::endl;
	return -1;
    }
    //Variables required for the position control
    float robot_x;
    float robot_y;
    float robot_a;
    float goal_x;
    float goal_y;
    float goal_a;
    std::string control_type = "";

    //Getting parameteres to set goal position and control type.
    n.param<std::string>("type", control_type, "diff"); //The default control type is for a differential base
    n.param<float>("a", goal_a, 0.0);                   //The default goal angle is 0.0 rad
    if(!n.getParam("x", goal_x))
    {
	std::cout << "Invalid x value." << std::endl;
	return -1;
    }
    if(!n.getParam("y", goal_y))
    {
	std::cout << "Invalid y value." << std::endl;
	return -1;
    }
    if(control_type != "diff"  && control_type != "omni")
    {
	std::cout << "Valid control types are \"omni\" and \"diff\"." << std::endl;
	return -1;
    }
    
    if(control_type == "omni")
	std::cout << "Executing position control for an OMNIDIRECTIONAL base with " << std::endl;
    else
	std::cout << "Executing position control for a DIFFERENTIAL base with " << std::endl;
    std::cout << "Goal X=" << goal_x << "\tGoal Y=" << goal_y << "\tGoal Angle=" << goal_a << std::endl;

    //Required subscriptors, publishers and messages
    ros::Publisher pub_cmd_vel = n.advertise<geometry_msgs::Twist>("/hardware/mobile_base/cmd_vel", 1);
    ros::Rate loop(10);
    tf::TransformListener tl;
    tf::StampedTransform t;
    tf::Quaternion q;
    geometry_msgs::Twist msg_cmd_vel;

    while(ros::ok())
    {
	//Instructions needed to get the current robot position
	tl.waitForTransform("map", "base_link", ros::Time::now(), ros::Duration(0.5));
	tl.lookupTransform("map", "base_link", ros::Time(0), t);
	robot_x = t.getOrigin().x();
	robot_y = t.getOrigin().y();
	q = t.getRotation();
	robot_a = atan2(q.z(), q.w())*2;

	/*
	 * TODO:
	 *
	 * Calculate the values of msg_cmd_vel according to the control type.
	 * (for omnidirectional or differential base)
	 * Declare more variables if necessary.
	 * The maximum linear speed must be 0.5 m/s
	 * The maximul angular speed must be 0.5 rad/s
	 * The robot must stop if it is at 0.1 m or less from the goal point
	 */

    float error_x = goal_x - robot_x;
    float error_y = goal_y - robot_y;
    float error_a = goal_a - robot_a;
    float k   = 1;
    float ka  = 1;
    float v_x = 0;
    float v_y = 0;
    float w   = 0;
    if(control_type == "omni"){
        float v_x_a = k * error_x;
        float v_y_a = k * error_y;
        v_x = v_x_a * cos(robot_a) + v_y_a * sin(robot_a);
        v_y = -v_x_a * sin(robot_a) + v_y_a * cos(robot_a);
        w   = ka * error_a;
        if (v_x > 0.5)  v_x = 0.5;
        if (v_x < -0.5) v_x = -0.5;
        if (v_y > 0.5) v_y = 0.5;
        if (v_y < -0.5) v_y = -0.5;
        if (w > 0.5) w = 0.5;
        if (w < -0.5) w = -0.5;
        
    }else{
        float v_max = 0.5;
        float w_max = 0.5;
        float alfa = 1;
        float beta = 1;
        error_a = atan2(error_y,error_x) - robot_a;
        if(error_a > M_PI)  error_a -= 2*M_PI;
        if(error_a < -M_PI) error_a += 2*M_PI;
        v_x = v_max * exp((-error_a*error_a)/(alfa));
        v_y = 0;
        w   = w_max * ((2/(1 + exp(-error_a/beta))) - 1);

        float dist = sqrt((error_x*error_x)+(error_y*error_y));
        if(dist < 0.2){
            v_x = 0;
            w   = 0;
        } 
    }

    msg_cmd_vel.linear.x = v_x;
    msg_cmd_vel.linear.y = v_y;
    msg_cmd_vel.angular.z = w;
	pub_cmd_vel.publish(msg_cmd_vel);
	ros::spinOnce();
	loop.sleep();
    }
    
    return 0;
    
}

